### PR TITLE
feat: implement multi-tenancy and group export 

### PR DESCRIPTION
**Pull request by carvantes**
This is it. This is the final merge of the `feat-smart-multitenancy` branch into mainline

Most changes have been already reviewed before in separate PRs. Since this is a separate branch from `mainline` the `feat-multitenancy` branch couldn't be merged as is. The commits were cherry picked and a few of them required very minor modifcations to "fit" in the `smart-develop` branch.

All integ tests passed on a test branch: https://github.com/awslabs/fhir-works-on-aws-deployment/actions/runs/1161939792 (except for the last merge job that was expected to fail)

You may want to focus on reviewing the last few commits after chore: upgrade dependencies: https://github.com/awslabs/fhir-works-on-aws-deployment/commit/62e750557bba33dd4c582afcff946e10ad4640a7.

Checklist:

* [x] Have you successfully deployed to an AWS account with your changes?
* [x] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: integration-tests/multitenancy.test.ts [hunk 1]
```diff
@@ -146,52 +146,54 @@ if (process.env.MULTI_TENANCY_ENABLED === 'true') {
             client = await getFhirClient('fhirUser user/*.*', true, { tenant: 'tenant1' });
         });
         test('requests without /tenant/<tenantId> in path should fail', async () => {
-            await expect(client.get(`${process.env.API_URL}/Patient`)).rejects.toMatchObject({
+            await expect(client.get(`${process.env.SMART_SERVICE_URL}/Patient`)).rejects.toMatchObject({
```
Comment (Bingjiling) [inline integration-tests/multitenancy.test.ts:149]: Where is the value set for SMART_SERVICE_URL?

File: .github/workflows/deploy-smart.yaml [hunk 2]
```diff
@@ -3,11 +3,51 @@ on:
...
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Use Node.js
+        uses: actions/setup-node@v1
+        with:
+          node-version: 12
+      - uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+      - name: Install dependencies
```
Comment (Bingjiling) [inline .github/workflows/deploy-smart.yaml]: Nit pick: Use the working directory keyword to replace `cd` and `cd..`. Something like: 

```
- name: Build Hapi validator
  run: mvn --batch-mode --update-snapshots --no-transfer-progress clean install
  working-directory: javaHapiValidatorLambda
```

File: .github/workflows/deploy-smart.yaml [hunk 3]
```diff
@@ -85,7 +137,7 @@ jobs:
           bundle install
       - name: Execute tests
         env:
-          SERVICE_URL: ${{ secrets.SMART_SERVICE_URL}}
+          SERVICE_URL: ${{ secrets[matrix.smartServiceURLSecretName]}}${{ matrix.serviceUrlSuffix }}
```
Comment (Bingjiling) [inline .github/workflows/deploy-smart.yaml:140]: Does this corresponds to `SMART_SERVICE_URL` used in [this file](https://github.com/awslabs/fhir-works-on-aws-deployment/blob/99594182336a8e0cda938643058ac3925d22dc31/integration-tests/multitenancy.test.ts)?

File: integration-tests/multitenancy.test.ts (hunk 1 above)
Comment (carvantes) [reply integration-tests/multitenancy.test.ts:149]: Here: https://github.com/awslabs/fhir-works-on-aws-deployment/blob/502c1a06155bd595cbb33821d1cb3b22429b1655/.github/workflows/deploy-smart.yaml#L198

File: .github/workflows/deploy-smart.yaml (hunk 3 above)
Comment (carvantes) [reply .github/workflows/deploy-smart.yaml:140]: yes

File: .github/workflows/deploy-smart.yaml (hunk 2 above)
Comment (carvantes) [reply .github/workflows/deploy-smart.yaml]: yeah, that's probably better. That part was not modified as part of this release and is only a minor improvement to readability so I'm not changing it now.

Plus, I'd rather change it first on mainline and then on smart mainline(as usuaI). I'll open a ticket to change it later.

File: .github/workflows/deploy-smart.yaml (hunk 3 above)
Comment (carvantes) [reply .github/workflows/deploy-smart.yaml:140]: nvm, actually no. this are the inferno tests.

the `SMART_SERVICE_URL` env variable used in integ tests is set here:
https://github.com/awslabs/fhir-works-on-aws-deployment/blob/3cc1379d7f0dfc31b7f35b3bd2599740c790a108/.github/workflows/deploy-smart.yaml#L198

It's a bit confusing to use different env var names for inferno tests and for integ tests, but here we are just following the existing naming convention.

Comment (Bingjiling) [reply .github/workflows/deploy-smart.yaml:140]: Got it! Is `${{ matrix.serviceUrlSuffix }}` needed for SMART_SERVICE_URL as well?

Comment (carvantes) [reply .github/workflows/deploy-smart.yaml:140]: No, integ tests know how to add `/tenant/<tenantId>` to the base url. Those tests deal with multiple tenants.

https://github.com/awslabs/fhir-works-on-aws-deployment/blob/3cc1379d7f0dfc31b7f35b3bd2599740c790a108/integration-tests/utils.ts#L117-L135

**Review by Bingjiling — APPROVED** [review]
Ship it!

**Pull request merged**